File: pact-jvm-matchers/src/main/scala/au/com/dius/pact/matchers/JsonBodyMatcher.scala
package au.com.dius.pact.matchers

import java.util.Collections

import au.com.dius.pact.matchers.util.JsonUtils
import au.com.dius.pact.model._
import au.com.dius.pact.model.matchingrules.MatchingRules
import com.typesafe.scalalogging.StrictLogging

import scala.collection.JavaConverters._

class JsonBodyMatcher extends BodyMatcher with StrictLogging {

  def matchBody(expected: HttpPart, actual: HttpPart, allowUnexpectedKeys: Boolean): java.util.List[BodyMismatch] = {
    (expected.getBody.getState, actual.getBody.getState) match {
      case (OptionalBody.State.MISSING, _) => Collections.emptyList()
      case (OptionalBody.State.NULL, OptionalBody.State.PRESENT) => Collections.singletonList(new BodyMismatch(null, actual.getBody.getValue,
        s"Expected empty body but received '${actual.getBody.getValue}'"))
      case (OptionalBody.State.NULL, _) => Collections.emptyList()
      case (_, OptionalBody.State.MISSING) => Collections.singletonList(new BodyMismatch(expected.getBody.getValue, null,
        s"Expected body '${expected.getBody.getValue}' but was missing"))
      case (_, _) => compare(Seq("$"), JsonUtils.parseJsonString(expected.getBody.getValue),
        JsonUtils.parseJsonString(actual.getBody.getValue), allowUnexpectedKeys, expected.getMatchingRules).asJava
    }
  }

  def valueOf(value: Any) = {
    value match {
      case s: String => s"'$value'"
      case null => "null"
      case _ => value.toString
    }
  }

  def typeOf(value: Any) = {
    if (value == null) {
      "Null"
    } else value match {
      case _: Map[Any, Any] =>
        "Map"
      case _: List[Any] =>
        "List"
      case _ =>
        value.getClass.getSimpleName
    }
  }

  def compare(path: Seq[String], expected: Any, actual: Any, allowUnexpectedKeys: Boolean,
              matchers: MatchingRules): List[BodyMismatch] = {
    (expected, actual) match {
      case (a: Map[String, Any], b: Map[String, Any]) => compareMaps(a, b, a, b, path, allowUnexpectedKeys, matchers)
      case (a: List[Any], b: List[Any]) => compareLists(a, b, a, b, path, allowUnexpectedKeys, matchers)
      case (_, _) =>
        if ((expected.isInstanceOf[Map[String, Any]] && !actual.isInstanceOf[Map[String, Any]]) ||
          (expected.isInstanceOf[List[Any]] && !actual.isInstanceOf[List[Any]])) {
          List(new BodyMismatch(expected, actual,
            s"Type mismatch: Expected ${typeOf(expected)} ${valueOf(expected)} but received ${typeOf(actual)} ${valueOf(actual)}",
            path.mkString("."), generateObjectDiff(expected, actual)))
        } else {
          compareValues(path, expected, actual, matchers)
        }
    }
  }

  private def generateObjectDiff(expected: Any, actual: Any) = {
    DiffUtilsKt.generateObjectDiff(JsonUtils.scalaObjectGraphToJavaObjectGraph(expected),
      JsonUtils.scalaObjectGraphToJavaObjectGraph(actual))
  }

  def compareListContent(expectedValues: List[Any], actualValues: List[Any], path: Seq[String],
                         allowUnexpectedKeys: Boolean, matchers: MatchingRules) = {
    var result = List[BodyMismatch]()
    for ((value, index) <- expectedValues.view.zipWithIndex) {
      if (index < actualValues.size) {
        result = result ++: compare(path :+ index.toString, value, actualValues(index), allowUnexpectedKeys, matchers)
      } else if (!Matchers.matcherDefined("body", path.asJava, matchers)) {
        result = result :+ new BodyMismatch(expectedValues, actualValues, s"Expected ${valueOf(value)} but was missing",
          path.mkString("."), generateObjectDiff(expectedValues, actualValues))
      }
    }
    result
  }

  def compareLists(expectedValues: List[Any], actualValues: List[Any], a: Any, b: Any, path: Seq[String],
                   allowUnexpectedKeys: Boolean, matchers: MatchingRules): List[BodyMismatch] = {
    if (Matchers.matcherDefined("body", path.asJava, matchers)) {
      logger.debug("compareLists: Matcher defined for path " + path)
      var result = Matchers.domatch[BodyMismatch](matchers, "body", path.asJava, expectedValues, actualValues, BodyMismatchFactory.INSTANCE).asScala.toList
      if (expectedValues.nonEmpty) {
        result = result ++ compareListContent(expectedValues.padTo(actualValues.length, expectedValues.head),
          actualValues, path, allowUnexpectedKeys, matchers)
      }
      result
    } else {
      if (expectedValues.isEmpty && actualValues.nonEmpty) {
        List(new BodyMismatch(a, b, s"Expected an empty List but received ${valueOf(actualValues)}",
          path.mkString("."), generateObjectDiff(expectedValues, actualValues)))
      } else {
        var result = compareListContent(expectedValues, actualValues, path, allowUnexpectedKeys, matchers)
        if (expectedValues.size != actualValues.size) {
          result = result :+ new BodyMismatch(a, b,
            s"Expected a List with ${expectedValues.size} elements but received ${actualValues.size} elements",
            path.mkString("."), generateObjectDiff(expectedValues, actualValues))
        }
        result
      }
    }
  }

  def compareMaps(expectedValues: Map[String, Any], actualValues: Map[String, Any], a: Any, b: Any, path: Seq[String],
                  allowUnexpectedKeys: Boolean, matchers: MatchingRules): List[BodyMismatch] = {
    if (expectedValues.isEmpty && actualValues.nonEmpty) {
      List(new BodyMismatch(a, b, s"Expected an empty Map but received ${valueOf(actualValues)}", path.mkString("."),
        generateObjectDiff(expectedValues, actualValues)))
    } else {
      var result = List[BodyMismatch]()
      if (allowUnexpectedKeys && expectedValues.size > actualValues.size) {
        result = result :+ new BodyMismatch(a, b,
          s"Expected a Map with at least ${expectedValues.size} elements but received ${actualValues.size} elements",
          path.mkString("."), generateObjectDiff(expectedValues, actualValues))
      } else if (!allowUnexpectedKeys && expectedValues.size != actualValues.size) {
        result = result :+ new BodyMismatch(a, b,
          s"Expected a Map with ${expectedValues.size} elements but received ${actualValues.size} elements",
          path.mkString("."), generateObjectDiff(expectedValues, actualValues))
      }
      if (Matchers.wildcardMatchingEnabled() && Matchers.wildcardMatcherDefined((path :+ "any").asJava, "body", matchers)) {
        actualValues.foreach(entry => {
          if (expectedValues.contains(entry._1)) {
            result = result ++: compare(path :+ entry._1, expectedValues.apply(entry._1), entry._2, allowUnexpectedKeys, matchers)
          } else if (!allowUnexpectedKeys) {
            result = result ++: compare(path :+ entry._1, expectedValues.values.head, entry._2, allowUnexpectedKeys, matchers)
          }
        })
      } else {
        expectedValues.foreach(entry => {
          if (actualValues.contains(entry._1)) {
            result = result ++: compare(path :+ entry._1, entry._2, actualValues(entry._1), allowUnexpectedKeys, matchers)
          } else {
            result = result :+ new BodyMismatch(a, b, s"Expected ${entry._1}=${valueOf(entry._2)} but was missing",
              path.mkString("."), generateObjectDiff(expectedValues, actualValues))
          }
        })
      }
      result
    }
  }

  def compareValues(path: Seq[String], expected: Any, actual: Any, matchers: MatchingRules): List[BodyMismatch] = {
    if (Matchers.matcherDefined("body", path.asJava, matchers)) {
      logger.debug("compareValues: Matcher defined for path " + path)
      Matchers.domatch[BodyMismatch](matchers, "body", path.asJava, expected, actual, BodyMismatchFactory.INSTANCE).asScala.toList
    } else {
      logger.debug("compareValues: No matcher defined for path " + path + ", using equality")
      if (expected == actual) {
        List[BodyMismatch]()
      } else {
        List(new BodyMismatch(expected, actual, s"Expected ${valueOf(expected)} but received ${valueOf(actual)}",
          path.mkString(".")))
      }
    }
  }
}

File: core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/JsonBodyMatcher.kt
package au.com.dius.pact.core.matchers

import au.com.dius.pact.core.matchers.Matchers.compareListContent
import au.com.dius.pact.core.matchers.Matchers.compareLists
import au.com.dius.pact.core.matchers.util.IndicesCombination
import au.com.dius.pact.core.matchers.util.LargestKeyValue
import au.com.dius.pact.core.matchers.util.memoizeFixed
import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.support.json.JsonParser
import au.com.dius.pact.core.support.json.JsonValue
import mu.KLogging
import java.math.BigInteger

object JsonBodyMatcher : BodyMatcher, KLogging() {

  override fun matchBody(
    expected: OptionalBody,
    actual: OptionalBody,
    context: MatchingContext
  ): BodyMatchResult {
    return when {
      expected.isMissing() -> BodyMatchResult(null, emptyList())
      expected.isEmpty() && actual.isEmpty() -> BodyMatchResult(null, emptyList())
      !expected.isEmpty() && actual.isEmpty() ->
        BodyMatchResult(null, listOf(BodyItemMatchResult("$",
          listOf(BodyMismatch(null, actual.valueAsString(), "Expected empty body but received '${actual.value}'")))))
      expected.isNull() && actual.isPresent() ->
        BodyMatchResult(null, listOf(BodyItemMatchResult("$",
          listOf(BodyMismatch(null, actual.valueAsString(), "Expected null body but received '${actual.value}'")))))
      expected.isNull() -> BodyMatchResult(null, emptyList())
      actual.isMissing() ->
        BodyMatchResult(null, listOf(BodyItemMatchResult("$",
          listOf(BodyMismatch(expected.valueAsString(), null, "Expected body '${expected.value}' but was missing")))))
      else -> {
        BodyMatchResult(null, compare(listOf("$"), JsonParser.parseString(expected.valueAsString()),
          JsonParser.parseString(actual.valueAsString()), context))
      }
    }
  }

  private fun valueOf(value: Any?) = when (value) {
    is String -> "'$value'"
    is JsonValue.StringValue -> "'${value.asString()}'"
    is JsonValue -> value.serialise()
    null -> "null"
    else -> value.toString()
  }

  private fun typeOf(value: Any?) = when {
    value is Map<*, *> -> "Map"
    value is JsonValue.Object -> "Map"
    value is List<*> -> "List"
    value is JsonValue.Array -> "List"
    value is JsonValue.Null -> "Null"
    value is JsonValue -> value.name
    value == null -> "Null"
    else -> value.javaClass.simpleName
  }

  fun compare(
    path: List<String>,
    expected: JsonValue,
    actual: JsonValue,
    context: MatchingContext
  ): List<BodyItemMatchResult> {
    return when {
      expected is JsonValue.Object && actual is JsonValue.Object ->
        compareMaps(expected, actual, path, context)
      expected is JsonValue.Array && actual is JsonValue.Array ->
        compareLists(expected, actual, path, context)
      expected is JsonValue.Object && actual !is JsonValue.Object ||
        expected is JsonValue.Array && actual !is JsonValue.Array ->
        listOf(BodyItemMatchResult(path.joinToString("."),
          listOf(BodyMismatch(expected, actual, "Type mismatch: Expected ${typeOf(expected)} " +
          "${valueOf(expected)} but received ${typeOf(actual)} ${valueOf(actual)}", path.joinToString("."),
          generateJsonDiff(expected, actual)))))
      else -> compareValues(path, expected, actual, context)
    }
  }

  private fun compareLists(
    expectedValues: JsonValue.Array,
    actualValues: JsonValue.Array,
    path: List<String>,
    context: MatchingContext
  ): List<BodyItemMatchResult> {
    val expectedList = expectedValues.values
    val actualList = actualValues.values
    val result = mutableListOf<BodyItemMatchResult>()
    val generateDiff = { generateJsonDiff(expectedValues, actualValues) }
    if (context.matcherDefined(path)) {
      logger.debug { "compareLists: Matcher defined for path $path" }
      for (matcher in context.selectBestMatcher(path).rules) {
        result.addAll(compareLists(path, matcher, expectedList, actualList, context, generateDiff) {
          p, expected, actual, context -> compare(p, expected, actual, context)
        })
      }
    } else {
      if (expectedList.isEmpty() && actualList.isNotEmpty()) {
        result.add(BodyItemMatchResult(path.joinToString("."),
          listOf(BodyMismatch(expectedValues, actualValues,
            "Expected an empty List but received ${valueOf(actualValues)}",
          path.joinToString("."), generateJsonDiff(expectedValues, actualValues)))))
      } else {
        result.addAll(compareListContent(expectedList, actualList, path, context, generateDiff) {
          p, expected, actual, context -> compare(p, expected, actual, context)
        })
        if (expectedList.size != actualList.size) {
          result.add(BodyItemMatchResult(path.joinToString("."), listOf(BodyMismatch(expectedList, actualList,
            "Expected a List with ${expectedList.size} elements but received ${actualList.size} elements",
            path.joinToString("."), generateJsonDiff(expectedValues, actualValues)))))
        }
      }
    }
    return result
  }

  private fun compareMaps(
    expectedValues: JsonValue.Object,
    actualValues: JsonValue.Object,
    path: List<String>,
    context: MatchingContext
  ): List<BodyItemMatchResult> {
    return if (expectedValues.isEmpty() && actualValues.isNotEmpty() && !context.allowUnexpectedKeys) {
      listOf(BodyItemMatchResult(path.joinToString("."),
        listOf(BodyMismatch(expectedValues, actualValues, "Expected an empty Map but received ${valueOf(actualValues)}",
        path.joinToString("."), generateJsonDiff(expectedValues, actualValues)))))
    } else {
      val result = mutableListOf<BodyItemMatchResult>()
      val generateDiff = { generateJsonDiff(expectedValues, actualValues) }
      val expectedEntries = expectedValues.entries
      val actualEntries = actualValues.entries
      if (context.matcherDefined(path)) {
        for (matcher in context.selectBestMatcher(path).rules) {
          result.addAll(Matchers.compareMaps(path, matcher, expectedEntries, actualEntries, context, generateDiff) {
            p, expected, actual -> compare(p, expected ?: JsonValue.Null, actual ?: JsonValue.Null, context)
          })
        }
      } else {
        result.addAll(context.matchKeys(path, expectedEntries, actualEntries, generateDiff))
        for ((key, value) in expectedEntries) {
          val p = path + key
          if (actualEntries.containsKey(key)) {
            result.addAll(compare(p, value, actualEntries[key]!!, context))
          }
        }
      }
      result
    }
  }

  private fun compareValues(
    path: List<String>,
    expected: JsonValue,
    actual: JsonValue,
    context: MatchingContext
  ): List<BodyItemMatchResult> {
    return if (context.matcherDefined(path)) {
      logger.debug { "compareValues: Matcher defined for path $path" }
      listOf(BodyItemMatchResult(path.joinToString("."),
        Matchers.domatch(context, path, expected, actual, BodyMismatchFactory)))
    } else {
      logger.debug { "compareValues: No matcher defined for path $path, using equality" }
      if (expected == actual) {
        listOf(BodyItemMatchResult(path.joinToString("."), emptyList()))
      } else {
        listOf(BodyItemMatchResult(path.joinToString("."),
          listOf(BodyMismatch(expected, actual, "Expected ${valueOf(expected)} (${typeOf(expected)}) " +
            "but received ${valueOf(actual)} (${typeOf(actual)})", path.joinToString(".")))))
      }
    }
  }
}
